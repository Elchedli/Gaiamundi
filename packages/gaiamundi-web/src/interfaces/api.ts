export interface ApiError {
  status?: number;
  name?: string;
  message?: string;
  description?: string;
  details?: {
    errors?: Array<{
      message: string;
      name: string;
      path: string[];
    }>;
  };
}

export interface ApiErrorResponse {
  data?: any;
  error?: ApiError;
}

export type ApiResponse<T> = T;

export type ApiPaginationMeta = {
  //
  start?: number;
  limit?: number;
  // OR
  page?: number;
  pageSize?: number;
  pageCount?: number;
  // Always
  total: number;
};

export type ApiData<T> = T & {
  id: number;
  created_at: string;
  updated_at: string;
};

export interface ApiDocument<T> {
  data: ApiData<T>;
  meta?: {
    availableLocales?: string[];
  };
}

export interface ApiCollection<T> {
  data: Array<ApiData<T>>;
  meta: {
    pagination: ApiPaginationMeta;
  };
}
