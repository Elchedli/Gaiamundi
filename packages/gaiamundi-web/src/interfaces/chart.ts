import {
  AreaChartIcon,
  BarChartIcon,
  BubbleChartIcon,
  ColumnChartIcon,
  LineChartIcon,
  LineColumnChartIcon,
  PieChartIcon,
  ScatterChartIcon,
} from 'components/Icons/Charts';
import { AnimationEasing, AnimationOptions } from 'eazychart-core';
import {
  AreaChart,
  AreaChartProps,
  BarChart,
  BarChartProps,
  BubbleChart,
  BubbleChartProps,
  ColumnChart,
  LineChart,
  LineChartProps,
  LineColumnChart,
  PieChart,
  PieChartProps,
  ScatterChart,
  ScatterChartProps,
} from 'eazychart-react';

type ChartConfigProps<T> = Omit<
  T,
  'scopedSlots' | 'animationOptions' | 'rawData'
> & {
  animationOptions: Omit<AnimationOptions, 'easing'> & {
    easing: AnimationEasing;
  };
};

export type ChartType =
  | 'column'
  | 'bar'
  | 'pie'
  | 'line'
  | 'scatter'
  | 'bubble'
  | 'area'
  | 'lineColumn';

export type ChartTypes = {
  [key in ChartType]: {
    Icon: React.FC;
    ChartComponent: React.FC;
  };
};

export type AllChartConfigProps = ChartConfigProps<BarChartProps> &
  ChartConfigProps<PieChartProps> &
  ChartConfigProps<LineChartProps> &
  ChartConfigProps<ScatterChartProps> &
  ChartConfigProps<BubbleChartProps> &
  ChartConfigProps<AreaChartProps>;

export type Chart = {
  name: string;
  type: ChartType;
  props: AllChartConfigProps;
};

export const CHART_TYPES: ChartTypes = {
  column: {
    Icon: ColumnChartIcon,
    ChartComponent: ColumnChart,
  },
  bar: {
    Icon: BarChartIcon,
    ChartComponent: BarChart,
  },
  pie: {
    Icon: PieChartIcon,
    ChartComponent: PieChart,
  },
  line: {
    Icon: LineChartIcon,
    ChartComponent: LineChart,
  },
  scatter: {
    Icon: ScatterChartIcon,
    ChartComponent: ScatterChart,
  },
  bubble: {
    Icon: BubbleChartIcon,
    ChartComponent: BubbleChart,
  },
  area: {
    Icon: AreaChartIcon,
    ChartComponent: AreaChart,
  },
  lineColumn: {
    Icon: LineColumnChartIcon,
    ChartComponent: LineColumnChart,
  },
};

export const getChartIconByType = (type: keyof typeof CHART_TYPES) => {
  return type in CHART_TYPES ? CHART_TYPES[type].Icon : null;
};

export type RawDatumType =
  | 'string'
  | 'number'
  | 'bigint'
  | 'boolean'
  | 'symbol'
  | 'undefined'
  | 'object'
  | 'function';

export type RawDatum = Record<string, any>;
