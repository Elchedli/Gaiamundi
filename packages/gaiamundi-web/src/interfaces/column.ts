export interface Column {
  name: string;
  sample: number;
  source: string;
  validity: string;
  isGeoCode: boolean;
}

export interface DatasetColumn extends Column {
  dataset: string;
}
