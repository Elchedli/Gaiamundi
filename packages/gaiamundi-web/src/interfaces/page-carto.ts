import { ApiData, ApiDocument } from './api';
import { GeoMapAttributes } from './geo-map';
import { User } from './user';

export interface Tags {
  id: number;
  name: string;
  type: string[];
}

export interface PageCartoAttributes {
  id?: string;
  name: string;
  owner?: ApiDocument<User>;
  map: GeoMapAttributes;
  html: HTMLElement;
  tags: Tags;
  cover?: MediaImage;
}

export type PageCarto = ApiData<PageCartoAttributes>;
