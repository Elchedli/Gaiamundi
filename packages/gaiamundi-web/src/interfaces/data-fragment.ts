import { ApiCollection, ApiDocument } from './api';
import { Column } from './column';
import { Dataset } from './dataset';

export interface DataFragmentStub {
  columns: number[];
  dataset: number;
  page_carto: number;
}

export interface DataFragment {
  columns: ApiCollection<Column>;
  dataset: ApiDocument<Dataset>;
}
