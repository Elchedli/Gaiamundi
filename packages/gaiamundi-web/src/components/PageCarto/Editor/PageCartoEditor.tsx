import { ApiData } from 'interfaces/api';
import { PageCarto } from 'interfaces/page-carto';
import { FC } from 'react';
import { PageCartoMap } from './PageCartoMap';
import { PageCartoPanels } from './PageCartoPanels';

type PageCartoEditorProps = {
  pageCarto: ApiData<PageCarto>;
};

export const PageCartoEditor: FC<PageCartoEditorProps> = ({ pageCarto }) => {
  const map = pageCarto.attributes.map.data;

  return (
    <div className="grid grid-cols-3 gap-4 h-full">
      <div className="col-span-2">
        <PageCartoMap map={map} />
      </div>
      <div className="col-span">
        <PageCartoPanels pageCarto={pageCarto} />
      </div>
    </div>
  );
};
