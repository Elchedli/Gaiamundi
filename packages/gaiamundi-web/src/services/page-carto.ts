import { PageCarto, PageCartoStub } from 'interfaces/page-carto';
import { ContentType, strapi } from './strapi';

export const createPageCarto = async (data: PageCartoStub) => {
  return await strapi.create<PageCartoStub>(ContentType.PAGE_CARTOS, data);
};

export const getLatestPageCartos = async (page: number, pageSize: number) => {
  return await strapi.get<PageCarto>(ContentType.PAGE_CARTOS, {
    populate: '*',
    sort: 'createdAt:desc',
    pagination: {
      page,
      pageSize,
    },
  });
};

export const getPageCartoById = async (id: number) => {
  return await strapi.getById<PageCarto>(ContentType.PAGE_CARTOS, id, {
    populate: {
      map: {
        populate: {
          geoJSON: true,
        },
      },
      owner: true,
      columns: {
        populate: {
          dataset: true,
        },
      },
    },
  });
};

export const uploadCsv = async (file: File) => {
  return await strapi.uploadFile(file, 'api::page-carto.page-carto');
};
