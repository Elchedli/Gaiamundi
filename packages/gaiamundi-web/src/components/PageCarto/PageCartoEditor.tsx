import { useQuery } from 'react-query';
import { FC, useCallback, useRef } from 'react';
import { MapChart, ResponsiveChartContainer } from 'eazychart-react';
import panzoom, { PanZoom } from 'panzoom';

import { ApiErrorAlert } from 'components/Alert/ApiErrorMessage';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { Alert } from 'components/Alert/Alert';
import { ApiData, ApiError } from 'interfaces/api';
import { PageCarto } from 'interfaces/page-carto';
import { getGeoJson } from 'services/geo-map';

import 'eazychart-css';
import ButtonGroup from 'components/Button/ButtonGroup';
import { Button } from 'components/Button/Button';
import { MinusIcon, PlusIcon } from '@heroicons/react/24/solid';

type PageCartoEditorProps = {
  pageCarto: ApiData<PageCarto>;
};

export const PageCartoEditor: FC<PageCartoEditorProps> = ({ pageCarto }) => {
  const elementRef = useRef<SVGSVGElement | null>(null);
  const panzoomRef = useRef<PanZoom | null>(null);
  const map = pageCarto.attributes.map.data;
  const geoJson = map.attributes.geoJSON.data;
  const geoJsonQuery = useQuery({
    queryKey: ['geoJSON', geoJson.id],
    queryFn: async () => getGeoJson(geoJson),
    // The query will not execute until the userId exists
    enabled: !!geoJson,
  });

  // Set up panzoom on mount, and dispose on unmount
  const panZoomCallback = useCallback((element: HTMLDivElement | null) => {
    if (element) {
      elementRef.current = element.querySelector('svg');
      if (elementRef.current) {
        panzoomRef.current = panzoom(elementRef.current, {
          zoomSpeed: 0.1,
          minZoom: 0.25,
          maxZoom: 4,
          smoothScroll: true,
        });
      }
    }
  }, []);

  const zoomIn = () => {
    if (elementRef.current) {
      const rect = elementRef.current.getBBox();
      const cx = rect.x + rect.width / 2;
      const cy = rect.y + rect.height / 2;
      panzoomRef.current?.smoothZoom(cx, cy, 1.25);
    }
  };

  const zoomOut = () => {
    if (elementRef.current) {
      const rect = elementRef.current.getBBox();
      const cx = rect.x + rect.width / 2;
      const cy = rect.y + rect.height / 2;
      panzoomRef.current?.smoothZoom(cx, cy, 0.75);
    }
  };

  if (geoJsonQuery.isLoading) {
    return <LoadingMessage label={'Chargement de la carte ...'} />;
  }

  if (geoJsonQuery.isError) {
    return <ApiErrorAlert error={geoJsonQuery.error as ApiError} />;
  }

  if (!geoJsonQuery.data) {
    return <Alert type="info">Impossible de charger la carte.</Alert>;
  }

  return (
    <div className="w-full h-full relative">
      <ButtonGroup pill={true} className="m-4 absolute z-50">
        <Button icon={PlusIcon} onClick={zoomIn} />
        <Button icon={MinusIcon} onClick={zoomOut} />
      </ButtonGroup>
      <div className="w-full h-full overflow-hidden" ref={panZoomCallback}>
        <ResponsiveChartContainer>
          <MapChart
            map={{
              geoDomainKey: 'admin',
              valueDomainKey: 'value',
              projectionType: 'geoMercator',
              stroke: 'black',
              fill: 'black',
            }}
            padding={{ top: 0, right: 50, bottom: 150, left: 50 }}
            colors={['white', 'pink', 'red']}
            geoJson={geoJsonQuery.data}
            data={geoJsonQuery.data.features.map(
              (feature: any, idx: number) => {
                return {
                  admin: feature.properties?.admin,
                  value: idx,
                };
              }
            )}
          />
        </ResponsiveChartContainer>
      </div>
    </div>
  );
};
