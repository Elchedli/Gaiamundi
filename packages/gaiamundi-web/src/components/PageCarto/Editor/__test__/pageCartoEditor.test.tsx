import { render } from '@testing-library/react';
import { ApiData } from 'interfaces/api';
import { PageCarto } from 'interfaces/page-carto';
import { QueryClient, QueryClientProvider } from 'react-query';
import {
  mockCoverApiDocument,
  mockDataFragmentsApiCollection,
  mockGeoMapApiDocument,
  mockOwnerApiDocument,
  mockTagsApiCollection,
} from 'utils/mocks/data';
import { PageCartoEditor } from '../PageCartoEditor';
describe('pageCartoEditor', () => {
  const mockPageCartoData: ApiData<PageCarto> = {
    id: 1,
    attributes: {
      name: 'Page Carto 1',
      owner: mockOwnerApiDocument,
      tags: mockTagsApiCollection,
      map: mockGeoMapApiDocument,
      cover: mockCoverApiDocument,
      html: '<p>Test HTML</p>',
      data_fragments: mockDataFragmentsApiCollection,
      created_at: '2022-02-07T00:00:00.000Z',
      updated_at: '2022-02-07T00:00:00.000Z',
    },
  };

  it('renders the correct children components', async () => {
    const queryClient = new QueryClient();
    const { getByText, container } = render(
      <QueryClientProvider client={queryClient}>
        <PageCartoEditor pageCarto={mockPageCartoData} />
      </QueryClientProvider>
    );
    expect(container).toMatchSnapshot();
    const pageMap = getByText('Chargement de la carte ...');
    expect(pageMap).toBeInTheDocument();
    const pagePanel = getByText('Donn√©es');
    expect(pagePanel).toBeInTheDocument();
  });
});
