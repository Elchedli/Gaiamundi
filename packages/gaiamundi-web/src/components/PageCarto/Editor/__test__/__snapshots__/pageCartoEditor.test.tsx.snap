// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pageCartoEditor renders the correct children components 1`] = `
<div>
  <div
    class="grid grid-cols-3 gap-4 h-full"
  >
    <div
      class="col-span-2"
    >
      <div
        class="flex grow justify-center items-center"
      >
        <div
          class="flex flex-col"
        >
          <svg
            class="m-auto"
            fill="#88CDD3"
            viewBox="0 0 40 40"
            width="200"
          >
            <path
              d="M20.201,5.169c-8.254,0-14.946,6.692-14.946,14.946c0,8.255,6.692,14.946,14.946,14.946 s14.946-6.691,14.946-14.946C35.146,11.861,28.455,5.169,20.201,5.169z M20.201,31.749c-6.425,0-11.634-5.208-11.634-11.634 c0-6.425,5.209-11.634,11.634-11.634c6.425,0,11.633,5.209,11.633,11.634C31.834,26.541,26.626,31.749,20.201,31.749z"
              opacity="0.5"
            />
            <path
              d="M26.013,10.047l1.654-2.866c-2.198-1.272-4.743-2.012-7.466-2.012h0v3.312h0 C22.32,8.481,24.301,9.057,26.013,10.047z"
            >
              <animatetransform
                attributeName="transform"
                attributeType="xml"
                dur="0.8s"
                from="0 20 20"
                repeatCount="indefinite"
                to="360 20 20"
                type="rotate"
              />
            </path>
          </svg>
          <h1
            class="text-center text-xl font-semibold"
          >
            Chargement de la carte ...
          </h1>
        </div>
      </div>
    </div>
    <div
      class="col-span"
    >
      <div
        class="h-full w-full p-2"
      >
        <div
          aria-orientation="horizontal"
          class="flex bg-blue-600 rounded-t-lg p-1 pb-0"
          role="tablist"
        >
          <button
            aria-controls="headlessui-tabs-panel-:r3:"
            aria-selected="true"
            class="rounded-t-lg p-2 text-sm font-medium leading-5 text-blue-700 ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 bg-white shadow"
            data-headlessui-state="selected"
            id="headlessui-tabs-tab-:r0:"
            role="tab"
            tabindex="0"
            type="button"
          >
            Hypertexte
          </button>
          <button
            aria-controls="headlessui-tabs-panel-:r4:"
            aria-selected="false"
            class="rounded-t-lg p-2 text-sm font-medium leading-5 text-blue-700 ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 text-blue-100 hover:bg-white/[0.12] hover:text-white"
            data-headlessui-state=""
            id="headlessui-tabs-tab-:r1:"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Données
          </button>
          <button
            aria-controls="headlessui-tabs-panel-:r5:"
            aria-selected="false"
            class="rounded-t-lg p-2 text-sm font-medium leading-5 text-blue-700 ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 text-blue-100 hover:bg-white/[0.12] hover:text-white"
            data-headlessui-state=""
            id="headlessui-tabs-tab-:r2:"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Indicateurs
          </button>
        </div>
        <div
          class="h-full rounded-b-lg border border-blue-700"
        >
          <div
            aria-labelledby="headlessui-tabs-tab-:r0:"
            class="px-2 pb-5 overflow-y-scroll"
            data-headlessui-state="selected"
            id="headlessui-tabs-panel-:r3:"
            role="tabpanel"
            style="max-height: calc(100vh - 64px * 2);"
            tabindex="0"
          >
            <div
              class="text-center "
            >
              en cours de développement
            </div>
          </div>
          <span
            aria-labelledby="headlessui-tabs-tab-:r1:"
            id="headlessui-tabs-panel-:r4:"
            role="tabpanel"
            style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
            tabindex="-1"
          />
          <span
            aria-labelledby="headlessui-tabs-tab-:r2:"
            id="headlessui-tabs-panel-:r5:"
            role="tabpanel"
            style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
